### PR TITLE
PyHSPlasma Animation Improvements

### PR DESCRIPTION
**Pull request by Hoikas**
Improves the animation bits of PyHSPlasma by adding correct duck-typing for many return values and fixes many of those pesky plCreatable casts. This doesn't fix every single problem in the land, but it hits all of the ones needed for exporting simple animations in Korman and then some :smile:.

### CODE REVIEW

File: Python/PRP/Animation/pyLeafController.cpp [hunk 1]
```diff
@@ -64,28 +68,27 @@ static PyObject* pyLeafController_setKeys(pyLeafController* self, PyObject* args
             PyErr_SetString(PyExc_TypeError, "setKeys expects a list of hsKeyFrames and an int");
             return NULL;
         }
-        ((pyKeyFrame*)itm)->fPyOwned = false;
```
Comment (zrax) [inline Python/PRP/Animation/pyLeafController.cpp]: Should this have been removed?  I would expect the controller to take ownership of the keys...

File: core/PRP/Animation/plLeafController.cpp [hunk 2]
```diff
@@ -441,7 +441,7 @@ void plLeafController::setKeys(const std::vector<hsKeyFrame*>& keys, unsigned in
     DeallocKeys();
     AllocKeys(keys.size(), type);
     for (size_t i=0; i<keys.size(); i++)
-        fKeys[i] = keys[i];
+        *fKeys[i] = *keys[i];
```
Comment (zrax) [inline core/PRP/Animation/plLeafController.cpp]: Good catch...  Looks like it also needs to be fixed for `setEaseControllers` though

File: Python/PRP/Animation/pyLeafController.cpp (hunk 1 above)
Comment (Hoikas) [inline Python/PRP/Animation/pyLeafController.cpp]: :wink:

File: core/PRP/Animation/plLeafController.cpp (hunk 2 above)
Comment (Hoikas) [inline core/PRP/Animation/plLeafController.cpp]: I think this is mirroring Cyan's implementation somewhat literally. IMO copying the keys is the correct solution in this case--the keys could be coming from another controller, adding the gotcha of one controller stealing another controller's owned keys. There's also the fPyOwned gotcha where we have a TypeError raised and some keys are leaked as a result.

Comment (zrax) [inline core/PRP/Animation/plLeafController.cpp]: Ah, good points.  But we still need `setEaseControllers` to get fixed as well :)

Comment (Hoikas) [inline core/PRP/Animation/plLeafController.cpp]: indeed

**Comment by zrax** [comment]
While you're at it (sorry :P), the keys and controllers for `pyLeafController` are somewhat inconsistent...  Calling `leaf.keys = foo` generates an error and recommends `leaf.setKeys(foo)`, but `leaf.easeControllers` is directly assignable...  It would be nice if these were consistent >.>

**Comment by Hoikas** [comment]
I was thinking the same thing. Do we want to get rid of setKeys entirely?

**Comment by zrax** [comment]
Yeah, I'd be fine with just using the `.keys =` and `.easeControllers =` methods, and removing `setKeys` from Python.

**Comment by Hoikas** [comment]
Updated to make pyLeafController.keys more awesome. I also implemented pyAGMasterMod and pyAGModifier while you were looking away :tongue:

File: Python/PRP/Animation/pyLeafController.cpp [hunk 3]
```diff
@@ -101,13 +82,44 @@ static PyObject* pyLeafController_getEaseControllers(pyLeafController* self, voi
 }
 
 static int pyLeafController_setType(pyLeafController* self, PyObject* value, void*) {
-    PyErr_SetString(PyExc_RuntimeError, "To set the key type, use setKeys");
+    PyErr_SetString(PyExc_RuntimeError, "To set the key type, use the keys setter");
```
Comment (zrax) [inline Python/PRP/Animation/pyLeafController.cpp]: Ah, that explains the inconsistency...

**Pull request merged**